### DISCUSSION TITLE
Include icons in mermaid

### DISCUSSION THREAD

**Discussion by francesco086** [squidfunk/mkdocs-material#4454] [Q&A]
Hi, I would like to use the fantastic [icons provided by mkdocs-material](https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/) in a mermaid diagram.
Unfortunately this does not work. It also does not work [the way suggested by mermaid](https://mermaid-js.github.io/mermaid/#/flowchart?id=basic-support-for-fontawesome).

Here the .md snippet:
```md
:fontawesome-brands-gitlab: <!-- just to demonstrate that I set everything colrrectly to visualize icons -->

```mermaid
flowchart TD
    B[<img src='https://iconscout.com/ms-icon-310x310.png' width='40' height='40' />]
    B-->C["fa:fa-car"]
    B-->D[fa:fa-car]
    B-->F(":heart:")
    B-->G(:heart:);
```

This is the output I get

![Screenshot 2022-10-04 at 15 51 10](https://user-images.githubusercontent.com/8321888/193837368-e4d9ccd9-7985-417c-9c5f-37cb564cbec7.png)

Any suggestion?

**Comment by squidfunk** [comment]
@facelessuser any ideas? A custom fence might do the trick, converting icons to SVGs, right? Some styling might be necessary.

**Reply by facelessuser** [reply]
1. Default FontAwesome as described in their documentation doesn't work for me, even when I enable `htmlLabels` which I've historically had not great results with anyways, but maybe it is better in recent versions 🤷🏻 .
2. I also see this open issue which makes it ambiguous as to whether the issue has been fixed or not: https://github.com/mermaid-js/mermaid/issues/1435.

I'm not sure about trying to bring in Material icons. Is it possible? Probably. Is it easily doable? That I don't know. Would I prefer their built-in support worked and defer to that? Probably 🙃.

**Reply by facelessuser** [reply]
There may be a way to get their built-in font awesome to work. I'm not sure what the requirements exactly are.  I did see this comment in one of their issues:

> For me it's working. Be sure to include the css fontawesome AND the fonts itself. Make sure fonts/ directory is ../ relative to css/font-awesome.min.css

🤷🏻

**Reply by squidfunk** [reply]
Font Awesome might work if you use the web font integration (which we don't ship).

**Reply by facelessuser** [reply]
Yep, I'm also not sure if the manual way in which we trigger rendering side-steps font-awesome support. I'm not sure what's supposed to trigger it, so... 🤷🏻

**Reply by squidfunk** [reply]
If the web font is used, the authors just has to define one of the [supported gylphs](https://fontawesome.com/v5/docs/desktop/reference-icons/use-glyphs) and set the appropriate `font-family` property for the node or label or whatever should receive the icon. It's the same as what we did for showing icons in our UI prior to the (largely superior) icon integration we have now, e.g.:

https://github.com/squidfunk/mkdocs-material/blob/57f51f3a3de35923a8b513abb169ba0db12c70fd/src/assets/stylesheets/fonts/_icon.scss#L65-L67

That was version 0.2.4 😅

**Reply by squidfunk** [reply]
So to summarize: no changes would be necessary to support web fonts, the authors would just be responsible of ensuring to set the appropriate custom CSS to override `font-family` on the labels he wants to use FontAwesome. It's all user-land.

**Reply by francesco086** [reply]
How does a user "ensure to set the appropriate custom CSS to override font-family on the labels he wants to use FontAwesome"?

**Reply by squidfunk** [reply]
Our [customization guide](https://squidfunk.github.io/mkdocs-material/customization/#additional-css) explains how to add custom CSS. I think you should follow FontAwesome's guide explaining how to set up the web font and then add the appropriate CSS rules on the Mermaid classes. How exactly? I can't say without fiddling with it myself.

**Reply by facelessuser** [reply]
Would custom CSS be seen in the shadow DOM?

**Reply by squidfunk** [reply]
I've got another idea: Since you can use `img` tags, you might just use Material for MkDocs SVGs from a CDN (i.e. from GitHub):
https://github.com/squidfunk/mkdocs-material/tree/master/material/.icons – You can then use the [privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#built-in-privacy-plugin) (optional) which will automatically download the icons from GitHub and replace them with links to local copies. Something like:

```
flowchart TD
    B[<img src='https://raw.githubusercontent.com/squidfunk/mkdocs-material/master/material/.icons/material/library.svg' width='40' height='40' />]
    B-->C["fa:fa-car"]
    B-->D[fa:fa-car]
    B-->F(":heart:")
    B-->G(:heart:);
```

To find the image URL for an icon, just use [our icon search](https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/#search) to find the icon and then right-click on it to copy the image URL. Voila.

> Would custom CSS be seen in the shadow DOM?

Hmm, good question. It might probably be necessary to rebuild the theme and adjust the styles here:
https://github.com/squidfunk/mkdocs-material/blob/master/src/assets/javascripts/components/content/code/mermaid/index.css

In any case, using icons as images (what they are) is likely to be the simplest approach.

**Reply by francesco086** [reply]
I am far from being a web developer, but I suppose one could write a javascript function that given the icon name (e.g. "library"), returns the HTML img object, correct?

And one could add this as an extra, following https://squidfunk.github.io/mkdocs-material/customization/#additional-javascript

Is that right? Just to make the thing a little more readable... :)

**Reply by squidfunk** [reply]
If you're not a web developer, I'd say go for what I suggested in https://github.com/squidfunk/mkdocs-material/discussions/4454#discussioncomment-3797191. Much simpler.
